### ISSUE TITLE
Disallow groups with the exact same name

### ISSUE THREAD

**Issue by aguzmant103** [privacy-scaling-explorations/bandada#280]

**Comment by NamanAg0502** [comment]
Hey @aguzmant103,

Is it alright if the off-chain group name is same as on-chain group name. 

- [ ] Yes
- [ ] No

**Comment by aguzmant103** [comment]
Updating description: "Warn users if a new group has the exact same name"
FYI @NamanAg0502 

This should apply for both onchain and offchain groups

**Comment by aguzmant103** [comment]
Interesting behavior: if I create an offchain group with the same name as another offchain group, the old group members get deleted and overwritten. It seems the groupID is based on the group name

@vplasencia @0xjei is this expected? Woudl the expected behavior then be to NOT allow an offchain group with the same name as another offchain group?

- i.e. I created group "test" with groupID "85465432599585866400468606240662" with members "4,5,6"
- i.e. I create a new group "test" with group ID "85465432599585866400468606240662" with members "1,2,3"
- i.e. I create a group "test" with groupID "85465432599585866400468606240662" with members "7,8,9" with different fingerprint
- i.e. I create a group "test" with groupID "85465432599585866400468606240662" with members "1,2,3,4" with different fingerprint
-I query the API and dashboard and I only see the latest
- The groupID is based on the group name

![image](https://github.com/bandada-infra/bandada/assets/67167307/26472a26-9804-4cfa-a7aa-a6fabc4efe60)

**Comment by aguzmant103** [comment]
The same doesnt'y seem to happen with onchain groups. If you try to create an onchain group that already exist with that name i.e. "test" and "test" it'll error out

![image](https://github.com/bandada-infra/bandada/assets/67167307/820945b0-5bbc-4a24-9eb4-74854aada9c2)

**Comment by 0xjei** [comment]
> Interesting behavior: if I create an offchain group with the same name as another offchain group, the old group gets overwritten. It seems the groupID is based on the group name
> 
> @vplasencia @0xjei is this expected? Woudl the expected behavior then be to NOT allow an offchain group with the same name as another offchain group?
> 
> * i.e. I created group "test" with groupID "85465432599585866400468606240662" with members "4,5,6"
> * i.e. I create a new group "test" with group ID "85465432599585866400468606240662" with members "1,2,3"
> * i.e. I create a group "test" with groupID "85465432599585866400468606240662" with members "7,8,9" with different fingerprint
> * i.e. I create a group "test" with groupID "85465432599585866400468606240662" with members "1,2,3,4" with different fingerprint
> * I query the API and dashboard and I only see the latest

Thank you @aguzmant103 for the detailed report. This is a strange and unexpected behaviour and should be addressed as soon as possible. In my opinion, groups with the same name SHOULD NOT exist as there will be identifier collisions leading to inconsistency and overlapping of groups. This problem can be solved by checking the ID of the new group against those already created.

**Comment by 0xjei** [comment]
> The same doesnt'y seem to happen with onchain groups. If you try to create an onchain group that already exist with that name i.e. "test" and "test" it'll error out
> 
> Even if I change the admin (account address) the error appears if there's a group with that name
> 
> * [ ]  Improve Error Message when two onchain groups have the same name
> 
> ![image](https://private-user-images.githubusercontent.com/67167307/321697411-820945b0-5bbc-4a24-9eb4-74854aada9c2.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTMzNjQxNDUsIm5iZiI6MTcxMzM2Mzg0NSwicGF0aCI6Ii82NzE2NzMwNy8zMjE2OTc0MTEtODIwOTQ1YjAtNWJiYy00YTI0LTllYjQtNzQ4NTRhYWRhOWMyLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA0MTclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNDE3VDE0MjQwNVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTUxNTkxYjkzNjMzNzc4OThiNGJjZjc2MzFlOGU5ODljNGVmMTJhYjE0NDhjNTM1NGZiYTRlODE0NzY3YmI2YjkmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.hk9K_XIYqXjRTdgNYiLk-C2AND30Omj4wvrAmYN8qJg)

Correct. I noticed that for on-chain groups you are calling the `semaphore.createGroup()` function directly and not the `bandadaAPI.createGroup()` on the bandada dashboard. So the primer probably has some mechanism to check if a group with the same identifier already exists on the on-chain mapping (contracts). Does this make sense? Let me know! @aguzmant103

**Comment by aguzmant103** [comment]
@0xjei shall this compare per adminID or across all adminIDs?

In other words, if I create group "BandadaDevs" and you want to create the same group with the same name:
i) shall we allow it? (I think yes)
ii) does this clash with current database implementation? (idk, if so then we have to refactor deeper?)

**Comment by 0xjei** [comment]
> @0xjei shall this compare per adminID or across all adminIDs?
> 
> In other words, if I create group "BandadaDevs" and you want to create the same group with the same name: i) shall we allow it? (I think yes) ii) does this clash with current database implementation? (idk, if so then we have to refactor deeper?)

i) OK, I'll try to explain myself better. If we aim to have an ecosystem of groups, it would make sense for the identifiers (IDs) and names of the groups to be unique. This is to avoid that if I want to pass a group to a person, I am forced to pass the ID instead of the name (imagine a scenario with X groups with the same name, it could lead to confusion or scam). Thus, the name would act as the first filter, while the ID would act as the second level (the opposite happens in the backend logic). Tell me if you understand what I mean.
(ii) In case we wanted to support the same names for different groups in the same Bandada instance (infra), yes, currently it could be done but only because of that bug I told you about above.

**Comment by aguzmant103** [comment]
After discussing in team meeting: makes sense to check per admin to prevent the creation of two identical group

**Comment by aguzmant103** [comment]
@0xjei I'm going to return this to the backlog as I won't have bandwidth for now

**Comment by 0xjei** [comment]
> @0xjei I'm going to return this to the backlog as I won't have bandwidth for now

yes, makes sense - we can punt this back after the planning meeting accordingly to priorities <3

**Comment by vplasencia** [comment]
Hey @aguzmant103! I've just updated the issue description and unassigned you from this issue to allow others to take it. However, feel free to reassign it to yourself if you'd like to work on it.

**Comment by SIDHARTH20K4** [comment]
I am applying to this issue via [OnlyDust platform](https://app.onlydust.com/p/bandada).

### My background and how it can be leveraged
I have built many smart contracts so I can work on this issue

**Comment by vplasencia** [comment]
Hey @SIDHARTH20K4 do you want me to assign this issue to you?

**Comment by GoSTEAN** [comment]
I am applying to this issue via [OnlyDust platform](https://app.onlydust.com/p/bandada).

### My background and how it can be leveraged
I have experience in Typescript and solidity. My skills in front-end development ensure efficient, user-friendly solutions, particularly when handling data validation.

### How I plan on tackling this issue
Fetch Existing Groups: On dashboard load, retrieve the list of both on-chain and off-chain groups created by the user.
Implement Name Check: When a user enters a group name, validate it against the existing group names to detect any duplicates.
Trigger Warning: Display a user-friendly warning if a duplicate is detected, preventing submission.
Optimize UX: Ensure smooth UX by integrating real-time validation and providing clear feedback.
Test Thoroughly: Ensure the feature works for both on-chain and off-chain groups and covers edge cases.

**Comment by Deepak2623** [comment]
I am applying to this issue via [OnlyDust platform](https://app.onlydust.com/p/bandada).

### My background and how it can be leveraged
.

### How I plan on tackling this issue
.

**Comment by vplasencia** [comment]
Hey @GoSTEAN would you like me to assign this issue to you?

**Comment by GoSTEAN** [comment]
@vplasencia yes you can assign it to me

**Comment by vplasencia** [comment]
Hey @GoSTEAN! I just assigned the issue to you. Let us know if you have any questions.
